crates/core/src/event/extend.rs

```diff
@@ -0,0 +1,242 @@
+//! Extend [events](Event) with additional fields.
+
+use super::*;
+use crate::hash::Hash;
+use crate::storage::BlockHeight;
+
+/// Provides event composition routines.
+pub trait ComposeEvent {
+    /// Compose an [event](Event) with new data.
+    fn compose<NEW>(self, data: NEW) -> CompositeEvent<NEW, Self>
```

Comment:
Why do we need this? It seems like we could just use the `extend` method instead.

Reply:
to preserve the type of the base events. it will come in handy in a latter pr